src/openvic-simulation/diplomacy/DiplomaticAction.cpp

```diff
@@ -0,0 +1,127 @@
+#include "DiplomaticAction.hpp"
+
+#include <memory>
+#include <string_view>
+
+#include "openvic-simulation/types/IdentifierRegistry.hpp"
+#include "openvic-simulation/utility/Logger.hpp"
+
+using namespace OpenVic;
+
+DiplomaticActionType::DiplomaticActionType(std::string_view new_identifier, DiplomaticActionType::Initializer&& initializer)
+	: HasIdentifier(new_identifier), commit_action_caller { std::move(initializer.commit) },
+	  allowed_to_commit { std::move(initializer.allowed) }, get_acceptance { std::move(initializer.get_acceptance) } {}
+
+
+DiplomaticActionManager::DiplomaticActionManager() {}
+
+bool DiplomaticActionManager::add_diplomatic_action(
+	std::string_view identifier, DiplomaticActionType::Initializer&& initializer
+) {
+	if (identifier.empty()) {
+		Logger::error("Invalid diplomatic action identifier - empty!");
+		return false;
+	}
+	return diplomatic_action_types.add_item({ identifier, std::move(initializer) });
+}
+
+DiplomaticActionTickCache DiplomaticActionManager::create_diplomatic_action_tick(
+	std::string_view identifier, Country& sender, Country& reciever, std::any context_data
+) {
+	auto type = diplomatic_action_types.get_item_by_identifier(identifier);
+
+	DiplomaticActionTickCache result { { sender, reciever, context_data }, type };
+	if ((result.allowed_to_commit = result.type->allowed_to_commit(result.argument))) {
+		result.acceptance = result.type->get_acceptance(result.argument);
+	}
+	return result;
+}
+
+bool DiplomaticActionManager::setup_diplomatic_actions() {
+	using Argument = DiplomaticActionType::Argument;
+
+	bool result = true;
+
+	result &= add_diplomatic_action(
+		"form_alliance",
```

Comment:
Should continuous and cancellable actions (like `form_alliance`, `request_military_access`, `give_military_access`, `war_subsidies`, etc.) be marked as such and have a corresponding `allowed_to_cancel` function? (e.g. cancelling given military access can't be done with the target's troops still in your land, unless I'm getting confused with EU4)

Reply:
Yeah they probably should.